event/buffered_bus.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package event
+
+import (
+	"sync/atomic"
+)
+
+type subscribeCommand *Subscription
+
+type unsubscribeCommand *Subscription
+
+type publishCommand Message
+
+type closeCommand struct{}
+
+// bufferedBus is a bus that uses a buffered channel to manage subscribers and publish messages.
+type bufferedBus struct {
+	// subID is incremented for each subscriber and used to set subscriber ids.
+	subID atomic.Uint64
+	// subs is a mapping of subscriber ids to subscriptions.
+	subs map[uint64]*Subscription
+	// events is a mapping of event names to subscriber ids.
+	events map[string]map[uint64]struct{}
+	// commandChannel manages all commands sent to this simpleChannel.
+	//
+	// It is important that all stuff gets sent through this single channel to ensure
+	// that the order of operations is preserved.
+	//
+	// WARNING: This does mean that non-event commands can block the database if the buffer
+	// size is breached (e.g. if many subscribe commands occupy the buffer).
+	commandChannel  chan any
+	eventBufferSize int
+	hasClosedChan   chan struct{}
+	isClosed        atomic.Bool
+}
+
+// NewBufferedBus creates a new event bus with the given commandBufferSize and
+// eventBufferSize.
+//
+// Should the buffers be filled, subsequent calls on this bus will block.
+func NewBufferedBus(commandBufferSize int, eventBufferSize int) Bus {
+	bus := bufferedBus{
+		subs:            make(map[uint64]*Subscription),
+		events:          make(map[string]map[uint64]struct{}),
+		commandChannel:  make(chan any, commandBufferSize),
+		hasClosedChan:   make(chan struct{}),
+		eventBufferSize: eventBufferSize,
+	}
+	go bus.handleChannel()
+	return &bus
+}
+
+func (b *bufferedBus) Publish(msg Message) {
+	if b.isClosed.Load() {
+		return
+	}
+	b.commandChannel <- publishCommand(msg)
+}
+
+func (b *bufferedBus) Subscribe(events ...string) (*Subscription, error) {
+	if b.isClosed.Load() {
+		return nil, ErrSubscribedToClosedChan
+	}
+	sub := &Subscription{
+		id:     b.subID.Add(1),
+		value:  make(chan Message, b.eventBufferSize),
+		events: events,
+	}
+	b.commandChannel <- subscribeCommand(sub)
+	return sub, nil
+}
+
+func (b *bufferedBus) Unsubscribe(sub *Subscription) {
+	if b.isClosed.Load() {
+		return
+	}
+	b.commandChannel <- unsubscribeCommand(sub)
+}
+
+func (b *bufferedBus) Close() {
+	if b.isClosed.Load() {
+		return
+	}
+	b.isClosed.Store(true)
```

Comment:
todo: these 2 ops are not dome atomically and might close twice. Please use compare-and-swap:
```go
if !b.isClosed.CompareAndSwap(false, true) {
    return
}
...
```

Check please if you have similar occurrences.

Reply:
good catch!